### DISCUSSION TITLE
FastAPI does not allow to send a "None" UploadFile

### DISCUSSION THREAD

**Discussion by phuchm** [fastapi/fastapi#12040] [Questions] [answered]
### Privileged issue

- [X] I'm @tiangolo or he asked me directly to create an issue here.

### Issue Content

I have a POST method with FastAPI which has 2 params with UploadFile type. But one of them is required and one of them is Optional. When I send required param with None value, I got 422 error relating to "Field required". But if I send Optional param with None value, it's acceptable.
When I use:
from fastapi import Request
And add:
request: Request as extra param, then log request.form(), I don't see Optional param.
It seems to FastAPI ignored it?

My code is below:
```
from fastapi import File, Request, UploadFile

from typing import Annatoted

app: FastAPI = FastAPI()

@app.post(/test)
async def test(
    request: Request,
    firstFile: Annatoted[UploadFile, File(...)],
    secondFile: Annatoted[UploadFile | None, File(...)] = None):
    logger.debug(await request.form())
```

**Comment by YuriiMotov** [comment]
You don't need to use ellipsis with `File` when you use `Annotated`.

To fix your code you should remove ellipsis and use `SkipJsonSchema` to fix openapi documentation:

```
from fastapi import FastAPI, File, Request, UploadFile
from typing import Annotated
from pydantic.json_schema import SkipJsonSchema

app: FastAPI = FastAPI()

@app.post("/test")
async def test(
    request: Request,
    firstFile: Annotated[UploadFile, File()],
    secondFile: Annotated[UploadFile | SkipJsonSchema[None], File()] = None
):
    logger.debug(await request.form())
```

**Reply by phuchm** [reply]
You mean that use SkipJsonSchema will make this Optional field will be error when sending None value?

**Reply by YuriiMotov** [reply]
`SkipJsonSchema` is needed to fix openapi schema. Without `SkipJsonSchema` the type of second parameter in Swagger will be string.

To make second parameter optional you just need to remove `...`

**Reply by phuchm** [reply]
What problem here is not openai document, ellipsis just a shortcut of my code. What I mean here is how "request" parameter can display "secondFile" if using requests.post() method to send a None value. Currently, if I send None value for firstFile, it will be error but when I send a valid firstFile but None value for secondFile, "request" parameter just display firstFile. I expect that "request" parameter can contain information of secondFile when I log request.form() or return error if None value is passed

**Reply by YuriiMotov** [reply]
Could you add the examples of requests you send?

**Reply by phuchm** [reply]
Here is my client code:
```
from requests import post

with open("a.txt", "rb") as f:
    response = post(
        url="http://localhost:8080/test",
        files=(
            ("firstFile", ("a.txt", f, "application/octet-stream")),
            ("secondFile", None)
        ),

        timeout=None
    )
```

And with server code as above, you will get a FormData object from logger.debug(await request.form()). This object only contains firstFile without secondFile while I expected secondFile will be inside this object with None value

**Reply by YuriiMotov** [reply]
You should use triple \`\`\` symbol to format multi-line code

**Reply by phuchm** [reply]
I'm sorry. I corrected it. My mistake when using mobile :)

**Reply by YuriiMotov** [reply, marked as answer]
I think the problem is that `requests` just ignores the file parameter that is `None` and sends the request without it.
You can use the code below to see the raw request, that is sent:

```
import requests

def pretty_print_POST(req):
    print('{}\n{}\r\n{}\r\n\r\n{}'.format(
        '-----------START-----------',
        req.method + ' ' + req.url,
        '\r\n'.join('{}: {}'.format(k, v) for k, v in req.headers.items()),
        req.body,
    ))

with open("a.txt", "rb") as f:
    req = requests.Request(
        'POST',
        "http://localhost:8080/test",
        files=(
            ("firstFile", ("a.txt", f, "application/octet-stream")),
            ("secondFile", None)
        ),

    )
    prepared = req.prepare()
    pretty_print_POST(prepared)

```

The output shows that only `firstFile` is sent:

```

-----------START-----------
POST http://localhost:8080/test
Content-Length: 1985
Content-Type: multipart/form-data; boundary=7b19875569a580e9c5efa930a9c8e060


--7b19875569a580e9c5efa930a9c8e060
Content-Disposition: form-data;
name="firstFile";
filename="a.txt"
Content-Type: application/octet-stream

...file content goes here...
--7b19875569a580e9c5efa930a9c8e060--

```

**Reply by phuchm** [reply]
Yes, I saw. So, it seems to be problem of requests library, not FastAPI. I will investigate more about requests. Thank you so much for your support!
